File: ios/Runner/BILSystemCryptoBridge.swift
import CryptoKit
import Flutter
import Foundation

/// AES-256-GCM backed exclusively by Apple CryptoKit.
final class BILSystemCryptoBridge {
  private static let channelName = "bil/system_crypto"
  private static let keyLength = 32
  private static let nonceLength = 12
  private static let tagLength = 16

  static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: channelName,
      binaryMessenger: registrar.messenger()
    )
    channel.setMethodCallHandler(handle)
  }

  private static func handle(
    _ call: FlutterMethodCall,
    result: @escaping FlutterResult
  ) {
    do {
      switch call.method {
      case "encryptAes256Gcm":
        let arguments = try requiredArguments(call)
        let keyData = try requiredData(arguments, "key")
        let plaintext = try requiredData(arguments, "plaintext")
        guard keyData.count == keyLength else { throw CryptoBridgeError.invalidInput }

        let sealed = try AES.GCM.seal(
          plaintext,
          using: SymmetricKey(data: keyData)
        )
        let nonce = sealed.nonce.withUnsafeBytes { Data($0) }
        guard nonce.count == nonceLength else { throw CryptoBridgeError.invalidOutput }
        let protected = sealed.ciphertext + sealed.tag
        result([
          "nonce": FlutterStandardTypedData(bytes: nonce),
          "protected": FlutterStandardTypedData(bytes: protected),
        ])

      case "decryptAes256Gcm":
        let arguments = try requiredArguments(call)
        let keyData = try requiredData(arguments, "key")
        let nonceData = try requiredData(arguments, "nonce")
        let protected = try requiredData(arguments, "protected")
        guard keyData.count == keyLength,
              nonceData.count == nonceLength,
              protected.count >= tagLength else {
          throw CryptoBridgeError.invalidInput
        }

        let split = protected.count - tagLength
        let sealed = try AES.GCM.SealedBox(
          nonce: AES.GCM.Nonce(data: nonceData),
          ciphertext: protected.prefix(split),
          tag: protected.suffix(tagLength)
        )
        let plaintext = try AES.GCM.open(
          sealed,
          using: SymmetricKey(data: keyData)
        )
        result(FlutterStandardTypedData(bytes: plaintext))

      default:
        result(FlutterMethodNotImplemented)
      }
    } catch {
      // Fail closed without returning key, plaintext or CryptoKit details.
      result(FlutterError(
        code: "system_crypto_failed",
        message: "System AES-GCM operation failed.",
        details: nil
      ))
    }
  }

  private static func requiredArguments(
    _ call: FlutterMethodCall
  ) throws -> [String: Any] {
    guard let arguments = call.arguments as? [String: Any] else {
      throw CryptoBridgeError.invalidInput
    }
    return arguments
  }

  private static func requiredData(
    _ arguments: [String: Any],
    _ name: String
  ) throws -> Data {
    guard let value = arguments[name] as? FlutterStandardTypedData else {
      throw CryptoBridgeError.invalidInput
    }
    return value.data
  }

  private enum CryptoBridgeError: Error {
    case invalidInput
    case invalidOutput
  }
}

File: android/app/src/main/kotlin/com/bilhealth/bodyintelligencelog/BILSystemCryptoBridge.kt
package com.bilhealth.bodyintelligencelog

import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/** AES-256-GCM backed exclusively by Android's installed JCA provider. */
class BILSystemCryptoBridge(messenger: BinaryMessenger) {
    private val channel = MethodChannel(messenger, CHANNEL)
    private val secureRandom = SecureRandom()

    init {
        channel.setMethodCallHandler(::handle)
    }

    private fun handle(call: MethodCall, result: MethodChannel.Result) {
        try {
            when (call.method) {
                "encryptAes256Gcm" -> encrypt(call, result)
                "decryptAes256Gcm" -> decrypt(call, result)
                else -> result.notImplemented()
            }
        } catch (_: Exception) {
            // Fail closed without returning key, plaintext or provider details.
            result.error(
                "system_crypto_failed",
                "System AES-GCM operation failed.",
                null,
            )
        }
    }

    private fun encrypt(call: MethodCall, result: MethodChannel.Result) {
        val keyBytes = requiredBytes(call, "key")
        val plaintext = requiredBytes(call, "plaintext")
        require(keyBytes.size == KEY_LENGTH_BYTES)

        val nonce = ByteArray(NONCE_LENGTH_BYTES).also(secureRandom::nextBytes)
        val cipher = Cipher.getInstance(TRANSFORMATION)
        cipher.init(
            Cipher.ENCRYPT_MODE,
            SecretKeySpec(keyBytes, "AES"),
            GCMParameterSpec(TAG_LENGTH_BITS, nonce),
        )
        // JCA returns ciphertext followed by the 128-bit GCM authentication tag.
        val protectedBytes = cipher.doFinal(plaintext)
        result.success(mapOf("nonce" to nonce, "protected" to protectedBytes))
    }

    private fun decrypt(call: MethodCall, result: MethodChannel.Result) {
        val keyBytes = requiredBytes(call, "key")
        val nonce = requiredBytes(call, "nonce")
        val protectedBytes = requiredBytes(call, "protected")
        require(keyBytes.size == KEY_LENGTH_BYTES)
        require(nonce.size == NONCE_LENGTH_BYTES)
        require(protectedBytes.size >= TAG_LENGTH_BYTES)

        val cipher = Cipher.getInstance(TRANSFORMATION)
        cipher.init(
            Cipher.DECRYPT_MODE,
            SecretKeySpec(keyBytes, "AES"),
            GCMParameterSpec(TAG_LENGTH_BITS, nonce),
        )
        result.success(cipher.doFinal(protectedBytes))
    }

    private fun requiredBytes(call: MethodCall, name: String): ByteArray =
        requireNotNull(call.argument<ByteArray>(name))

    companion object {
        private const val CHANNEL = "bil/system_crypto"
        private const val TRANSFORMATION = "AES/GCM/NoPadding"
        private const val KEY_LENGTH_BYTES = 32
        private const val NONCE_LENGTH_BYTES = 12
        private const val TAG_LENGTH_BYTES = 16
        private const val TAG_LENGTH_BITS = TAG_LENGTH_BYTES * 8
    }
}
